packages/core/src/routes/experience/classes/helpers.ts

```diff
@@ -171,6 +178,59 @@ export const mergeUserMfaVerifications = (
   return [...userMfaVerifications, ...newMfaVerifications];
 };
 
+/**
+ * Filter out backup codes mfa verifications that have been used
+ */
+const filterOutEmptyBackupCodes = (
+  mfaVerifications: User['mfaVerifications']
+): User['mfaVerifications'] =>
+  mfaVerifications.filter((mfa) => {
+    if (mfa.type === MfaFactor.BackupCode) {
+      return mfa.codes.some((code) => !code.usedAt);
+    }
+    return true;
+  });
+
+/**
+ * Get all enabled MFA verifications for a user (stored + implicit)
+ * @param mfaSettings - MFA settings from sign-in experience
+ * @param user - User object with mfaVerifications and profile data
+ * @param userProfile - Optional profile override (for interaction contexts)
+ * @returns Array of all enabled MFA verifications
+ */
+export const getAllUserEnabledVerifications = (
+  mfaSettings: Mfa,
+  user: User,
+  userProfile?: { primaryEmail?: string }
+): MfaVerification[] => {
+  const storedVerifications = filterOutEmptyBackupCodes(user.mfaVerifications).filter(
+    (verification) => mfaSettings.factors.includes(verification.type)
+  );
+
+  if (!EnvSet.values.isDevFeaturesEnabled) {
+    return storedVerifications;
+  }
+
+  const profile = userProfile ?? {
+    primaryEmail: user.primaryEmail ?? undefined,
+  };
```

Comment:
What if the userProfile does not contain `primaryEmail` but `user.primaryEmail` exists?